### DISCUSSION TITLE
Would it be possible to get Alltalk to read the contents of the clipboard?

### DISCUSSION THREAD

**Discussion by djmeowth9** [erew123/alltalk_tts#272] [Q&A]
Hi! Alltalk is amazing. As someone with reading difficulties who is dissatisfied with the usual robotic-sounding TTS voices, thanks so much for developing it. 

I have a question about a rather niche use case for Alltalk. I play several very text-heavy games with minimal voice acting. I've used Alltalk to clone some of the few voice samples in these games, and created voices that can read the rest of the unvoiced dialogue (which I copy from the game using ctrl+c). It's awesome! But to do this I'm constantly tabbing between the game and the TTS generator, which is a bit cumbersome and pulls you out of the game (quite literally). Could there perhaps be a way for Alltalk to monitor the Windows clipboard, and when new text is copied to it, voice the text using the currently selected character voice in the TTS generator? If not, how difficult would it be to implement this feature? 

Thank you again!

**Comment by RenNagasaki** [comment]
I guess that's not really something alltalk should/could easily and cleanly handle.  I'd recommend writing a small program which consumes allTalk as a service. If I may ask which games do you play? For several games theres already good solutions out there. For one I'm currently writing my own.

**Reply by RenNagasaki** [reply]
@djmeowth9 forgot to tag you.

**Reply by djmeowth9** [reply]
@RenNagasaki 

That's fair, I know it's probably beyond the scope of the project. 

I mainly play World of Warcraft and Final Fantasy XIV. They have reams and reams and reams of unvoiced text which is a big strain for me to read, but I enjoy the stories, so up until now I've been making do with some text to speech addons for both, using Amazon's Neural Polly voices. They do the job of reading the text, but after cloning some voices from each game, it really made me want some kind of TTS solution that can use those cloned voices. It's much more immersive to hear the NPCs speak to you in their actual voice, or at least, an appropriate-sounding voice for the world of the game. 

I wonder if it's possible to create new 'SAPI 5' voices that Windows can use for TTS, out of voices generated by Alltalk or similar projects?

**Reply by RenNagasaki** [reply]
@djmeowth9 If I remember correctly WoW has one Addon where the whole game got Voice Acted. Lemme check if I can find it.

For FF14 You're in Luck. That's the Game I'm currently writing a plugin for. 🤣
Heres the Link: [https://github.com/RenNagasaki/Echokraut](https://github.com/RenNagasaki/Echokraut)
It's still unfinished but I can always use more testers.
The Description isn't up 2 date theres far more features already than written there. I'll update that tomorrow.
You should be able to simply install it, setup the alltalk connection, maybe rename some voices, and then you can play on!

@erew123  If you don't like me linking that here, just tell me and I'll remove it 😊

**Reply by RenNagasaki** [reply]
@djmeowth9 This would be the WoW Addon I meant. https://www.reddit.com/r/classicwow/comments/122avx7/voiceover_addon_full_release_revolutionize_your/

**Reply by djmeowth9** [reply]
@RenNagasaki That FF14 plugin looks awesome! I'll give it a shot as soon as I can! As for the WoW addon, that's what made me fall down this rabbit hole of trying to game with voice cloned dialogue to begin with! Unfortunately it only voices dialogue for the classic version of the original 2004 release, so anything introduced afterwards is still unvoiced. It also doesn't really utilize the user's self-hosted TTS, but rather, uses premade voice files that the addon developer made using elevenlabs. Something like your FF14 plugin would be great for WoW, but I'm not sure if that's possible given the differences between the games. Definitely gonna try out Echokraut though!

**Reply by RenNagasaki** [reply]
Oh, WoW is much more open in regards to addons. It should be totally possible to do this. It just needs someone to code it. 😆

**Comment by erew123** [comment]
Hi @RenNagasaki and @djmeowth9 

First off @RenNagasaki - No problems sharing links of the kind you did, thats perfectly fine by me :) And thanks for chipping in on the conversation too!

@djmeowth9 If I was to do something like that...... then I guess a rough shot at the code would look something.... something.......like this:

https://github.com/erew123/alltalk_tts/tree/alltalkbeta/system/cliptotts

![image](https://github.com/user-attachments/assets/92c757ae-6eb6-440b-a809-2c904d43c3b2)

![image](https://github.com/user-attachments/assets/59c02dbb-ea7e-46e4-871a-48a0702842ca)

![image](https://github.com/user-attachments/assets/d226cb6e-47ec-467e-8272-1a6e056462b6)

But it would be a rough bit of code, no logic or real extra controls, just the basics. However, it would save having to copy/paste text all the time! :)

Oh, and it would hang if you DONT have the icon.png file in the same folder.....

Thanks

**Reply by djmeowth9** [reply]
@erew123 

Wow! Thanks so much for putting this together! 

Running into two issues with it, though - firstly, once I open the settings menu for the clipboard-to-tts system tray icon, exiting it in any way causes it to close. This makes it impossible to change the voice since clicking 'save' closes the options menu as well as the clip-board-to-tts process/program entirely. If I run clipboard-to-tts again, the voice is once again female_01.wav even if I changed it.

Next thing I tried was copying and renaming one of my cloned voices to female_01.wav since that's the default in the settings. This gives me the error "TTS generation failed with status code: 422" in the cmd window. Do I need specific settings enabled in the Alltalk Startup Settings?

Sorry if these are extremely obvious answers, I'm very much not a programmer!

**Reply by erew123** [reply]
@djmeowth9 Well I did promise you some rough code, so at least I delivered on that! ;)

Its not something Ill spend a lot of time on, but I've done a few quick updates which you should find Ive added a few extra features and all should be good now.

Thanks

**Reply by erew123** [reply]
@djmeowth9 Obviously, just download the updated version and use it the same way!

Thanks
